kv/dist_sender.go

```diff
@@ -326,28 +329,74 @@ func (ds *DistSender) getRangeDescriptor(key proto.Key) ([]proto.RangeDescriptor
 	return ds.internalRangeLookup(metadataKey, desc)
 }
 
-// sendRPC sends one or more RPCs to replicas from the supplied
-// proto.Replica slice. First, replicas which have gossiped addresses
-// are corralled and then sent via rpc.Send, with requirement that one
-// RPC to a server must succeed. Returns an RPC error if the request
-// could not be sent. Note that the reply may contain a higher level
-// error and must be checked in addition to the RPC error.
-func (ds *DistSender) sendRPC(desc *proto.RangeDescriptor, method string, args proto.Request, reply proto.Response) error {
+// sendRPC sends one or more RPCs to replicas from the supplied proto.Replica
+// slice. First, replicas which have gossiped addresses are corralled (and
+// rearranged depending on proximity and whether the request needs to go to a
+// leader) and then sent via rpc.Send, with requirement that one RPC to a
+// server must succeed. Returns an RPC error if the request could not be sent.
+// Note that the reply may contain a higher level error and must be checked in
+// addition to the RPC error.
+func (ds *DistSender) sendRPC(desc *proto.RangeDescriptor, method string,
+	args proto.Request, reply proto.Response) error {
 	if len(desc.Replicas) == 0 {
 		return util.Errorf("%s: replicas set is empty", method)
 	}
 
+	// Unless we know better, send the RPCs randomly.
+	order := rpc.OrderingPolicy(rpc.OrderRandom)
+	// Try to look up our own NodeDescriptor and sort replicas by proximity.
```

Comment:
s/proximity/ordered attribute affinity, which we treat as a stand-in for proximity/